engine/src/flutter/impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -50,496 +52,581 @@ class PositionWriter {
-        cap_proc(vtx_builder, polyline.GetPoint(contour_start_point_i),
-                 cap_offset, scale, /*reverse=*/true);
-      }
+  void RecordConic(Point p1, Point cp, Point p2, Scalar weight) override {
+    RecordCurve<PathTessellator::Conic>({p1, cp, p2, weight});
+  }
 
-      for (size_t contour_component_i = 0;
-           contour_component_i < contour.components.size();
-           contour_component_i++) {
-        const Path::PolylineContour::Component& component =
-            contour.components[contour_component_i];
-        bool is_last_component =
-            contour_component_i == contour.components.size() - 1;
-
-        size_t component_start_index = component.component_start_index;
-        size_t component_end_index =
-            is_last_component ? contour_end_point_i - 1
-                              : contour.components[contour_component_i + 1]
-                                    .component_start_index;
-        if (component.is_curve) {
-          AddVerticesForCurveComponent(
-              vtx_builder, component_start_index, component_end_index,
-              contour_start_point_i, contour_end_point_i, contour);
-        } else {
-          AddVerticesForLinearComponent(
-              vtx_builder, component_start_index, component_end_index,
-              contour_start_point_i, contour_end_point_i, contour);
+  void RecordCubic(Point p1, Point cp1, Point cp2, Point p2) override {
+    RecordCurve<PathTessellator::Cubic>({p1, cp1, cp2, p2});
+  }
+
+  template <typename Curve>
+  inline void RecordCurve(const Curve& curve) {
+    auto start_direction = curve.GetStartDirection();
+    auto end_direction = curve.GetEndDirection();
+
+    // The Prune receiver should have eliminated any empty curves, so any
+    // curve we see should have both start and end direction.
+    FML_DCHECK(start_direction.has_value() && end_direction.has_value());
+
+    // In order to keep the compiler/lint happy we check for values anyway.
+    if (start_direction.has_value() && end_direction.has_value()) {
+      // We now know the curve cannot be degenerate
+      auto start_perpendicular = ComputePerpendicular(-start_direction.value());
+      auto end_perpendicular = ComputePerpendicular(end_direction.value());
+
+      // We join the previous segment to this one with a normal join
+      // The join will append the perpendicular at the start of this
+      // curve as well.
+      HandlePreviousJoin(start_perpendicular);
+
+      Scalar count =
+          std::ceilf(curve.SubdivisionCount(scale_ * half_stroke_width_));
+
+      Point prev = curve.p1;
+      auto prev_perpendicular = start_perpendicular;
+
+      // Handle all intermediate curve points up to but not including the end
+      for (int i = 1; i < count; i++) {
+        Point cur = curve.Solve(i / count);
+        auto cur_perpendicular = ComputePerpendicular(prev, cur);
+        if (prev_perpendicular.GetAlignment(cur_perpendicular) <
+            trigs_[1].cos) {
```

Comment:
How do we know `trigs_[1]` is valid at this point?

Reply:
I updated the class comments on Trigs to specify its minimum configurations. I also added an FML_DCHECK in the constructor and I'll try to clarify in the comments.

I *could* copy this value into a separate field named to indicate its intention perhaps. Or use an accessor that would be inlined to return the same value.

Reply:
Comments added on Trigs class and DCHECKS added at the start of the process.